### PR TITLE
issue78

### LINKED ISSUE #78: setup open source rover hardware interface

### PR DESCRIPTION
**Pull request by yakobi99**
hardware interface implementation for ros_control. cmd, pos, vel & eff moved to protected. roboclaw class inherited from roboclaw. implemented drive methods for motors, taking velocity commands. protobot.h read() and write() changed to Read() and Write(), due to conflicts with unistd.h. Read() not fully implemented, but I wanna test it on the hardware first to make sure I', sending the checksums are in the right order

### CODE REVIEW

File: src/protobot_control/src/protobot.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+protobot::protobot()
+{
+    ROS_INFO("Registering ros_control handlers");
+    registerStateHandlers();
+    registerJointVelocityHandlers();
+
+    clock_gettime(CLOCK_MONOTONIC, &last_time);
+}
+
+void protobot::registerStateHandlers() {
```
Comment (ashtencodes) [inline src/protobot_control/src/protobot.cpp:16]: what urdf are these joints for? there only seems to be two joints here

**Review by ashtencodes — APPROVED** [review]
everything looks good, just curious about the defined joints and how easy it is to integrate into competition hchassis

File: src/protobot_control/src/protobot.cpp (hunk 1 above)
Comment (yakobi99) [reply src/protobot_control/src/protobot.cpp:16]: still working on that stuff

**Comment by bretwitt** [comment]
> everything looks good, just curious about the defined joints and how easy it is to integrate into competition hchassis

the joint state handlers are for the old URDF, there should be six

File: src/protobot_control/src/protobot.cpp (hunk 1 above)
Comment (bretwitt) [reply src/protobot_control/src/protobot.cpp:16]: update state handlers for 6 joints

File: src/protobot_control/src/protobot.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+//      {
+//        ROS_INFO_STREAM("READING");
+//        ROS_INFO("CMD TO RIGHT_WHEEL_JOINT %f", cmd[0]);
+//        ROS_INFO("CMD TO LEFT_WHEEL_JOINT %f", cmd[1]);
+//      }
+    double right_vel = cmd[0];
+    double left_vel = cmd[1];
+}
+*/
+
```
Comment (bretwitt) [inline src/protobot_control/src/protobot.cpp:72]: delete zombie code

File: src/protobot_control/src/roboclaw.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#include<fcntl.h>
+#include<errno.h>
+#include<termios.h>
+#include<unistd.h>
+#include<sys/ioctl.h>
+#include<sys/time.h>
+#include<assert.h>
+#include "protobot.h"
+
+class roboclaw : public protobot {
```
Comment (bretwitt) [inline src/protobot_control/src/roboclaw.h:21]: shouldn't inherit from protobot, you can use composition and hook/call roboclaw functionality in the control loop in main()

**Review by bretwitt — CHANGES_REQUESTED** [review]
PR's for merging functional tested code smh

**Pull request closed without merging**